src/game/UINationSelectDlg.cpp

```diff
@@ -333,3 +216,104 @@ void CUINationSelectDlg::Render() {
         CN3Base::s_lpD3DDev->SetTextureStageState(0, D3DTSS_ALPHAARG2, dwTexAlpha2);
     }
 }
+
+bool CUINationSelectDlg::ReceiveMessage(CN3UIBase * pSender, DWORD dwMsg) {
+    if (!pSender) {
+        return false;
+    }
+
+    if (!m_bTransitionActive && dwMsg == UIMSG_BUTTON_CLICK) {
+        if (pSender == (CN3UIBase *)m_pBtnCloseKa || pSender == (CN3UIBase *)m_pBtnCloseEl) {
+            CGameProcedure::s_pSocket->Disconnect();
+            CGameProcedure::ProcActiveSet((CGameProcedure *)CGameProcedure::s_pProcLogIn);
+            CGameProcedure::s_pProcLogIn->m_bLogIn = true;
```

Comment:
Curious question; why do we set here `m_bLogIn` to `true` when exiting the nation select procedure? Especially after exiting means that we're logged out. I suppose to indicate that the user is passed through the authentication stage if they are at the nation select procedure? In that case, maybe we should set it from the login procedure instead when we receive a successful authentication message from the server.

Reply:
Good question. It's how it's done officially in the v1264 client:
```c
  if ( !this->m_bTransitionActive && a_dwMsg == 1 )
  {
    if ( a_pSender == (CN3UIBase *)this->m_pBtnCloseKa )
    {
LABEL_6:
      CAPISocket::Disconnect(s_pSocket);
      CGameProcedure::ProcActiveSet((int)s_pProcLogIn);
      BYTE1(s_pProcLogIn->m_bLogIn) = 1;
      return 1;
    }
```

I was looking for references in the existing code and `m_bLogIn` is nowhere to be used on the login procedure level apart from assigning true or false. So I spend some time checking the assembly and the only place it was used is in `CUILogIn::ReceiveMessage`:
```c
....
  if ( a_dwMsg != 0x1000 )                      // UIMSG_EDIT_RETURN
  {
    if ( a_dwMsg == 0x200000 && a_pSender == (CN3UIBase *)this->m_ListServerInfos.heap )// UIMSG_LIST_SELCHANGE
      CUILogIn::ServerInfoUpdate(this);
    return 1;
  }
  v12 = s_pProcLogIn;
  if ( BYTE1(s_pProcLogIn->m_bLogIn) || LOBYTE(this->m_bLogIn) || !this->m_pEdit_id || !this->m_pEdit_pw )
  {
    if ( !LOBYTE(this->m_bOpenningNow) )
    {
      m_pGroup_ServerList = this->m_pGroup_ServerList;
      if ( m_pGroup_ServerList )
      {
        if ( LOBYTE(m_pGroup_ServerList->CN3UIBase.m_bVisible) )
        {
          pMsgBox = s_pMsgBoxMgr->m_pMsgBoxLatestRef;
          if ( !pMsgBox || !LOBYTE(pMsgBox->CN3UIBase.m_bVisible) )
          {
            CGameProcLogIn::ConnectToGameServer(s_pProcLogIn);
            v12 = s_pProcLogIn;
          }
          BYTE1(v12->m_bLogIn) = 0;
        }
      }
    }
    return 1;
  }
  m_pMsgBoxLatestRef = s_pMsgBoxMgr->m_pMsgBoxLatestRef;
  if ( !m_pMsgBoxLatestRef || !LOBYTE(m_pMsgBoxLatestRef->CN3UIBase.m_bVisible) )
    goto LABEL_7;
  return 1;
}
```

I suppose it's used to ensure globally (considering the static variable of the procedure is assigned) to not send login packets multiple times.
However I was not able to reproduce, since there is already existing check in code:
```c
    } else if (dwMsg == UIMSG_EDIT_RETURN) {
        if (!m_bLogIn && m_pEdit_id && m_pEdit_pw) {
            CN3UIBase * pMsgBox = CGameProcedure::s_pMsgBoxMgr->GetFocusMsgBox();
            if (!(pMsgBox && pMsgBox->IsVisible())) {
                CGameProcedure::s_pProcLogIn->MsgSend_AccountLogIn(LIC_KNIGHTONLINE);
            }
        }
    }
```

Long story short. I assign it there just to keep it as in the official client, but still unsure what scenario it covers.

Reply:
Makes sense. Many thanks for the detailed explanation and for spending the time checking this.